### ISSUE TITLE
Tip: how to make it summarize mid-tail languages, e.g. Polish

### ISSUE THREAD

**Issue by Manamama** [miso-belica/sumy#204]
### Problem:
The sumy module uses the nltk package for stemming and stop words, but nltk does not support e.g. the Polish language out of the box.
#### Stop words: 
Download the Polish stop words file from e.g. [here](https://github.com/stopwords-iso/stopwords-pl/raw/master/stopwords-pl.txt), rename it to ```polish.txt```, and place it in the sumy stop words directory (```~/.local/lib/python3.10/site-packages/sumy/data/stopwords/polish.txt```).
#### Stemming: 
Use the [pystempel package](https://github.com/dzieciou/pystempel/), which provides a stemmer for the Polish language. Here’s the code:

```
class CallableStemmer:
    def __init__(self, stemmer):
        self.stemmer = stemmer

    def __call__(self, word):
        return self.stemmer.stem(word)

def get_stemmer(language):
    if language == 'pol':
        # Create a StempelStemmer object for Polish
        stemmer_obj = StempelStemmer.default()
        # Wrap it in a CallableStemmer
        return CallableStemmer(stemmer_obj)
    else:
        # For non-Polish languages, use the original Stemmer
        return Stemmer(language)

``` 

Then in this section:

```
def handle_arguments(args, default_input_stream=sys.stdin):
    # ... (other code) ...

    language = args["--language"]
    if args["--stopwords"]:
        stop_words = read_stop_words(args["--stopwords"])
    else:
        stop_words = get_stop_words(language)

    parser = parser(document_content, Tokenizer(language))
    stemmer = get_stemmer(language)

    # ... (other code) ...
```
This way, if the language is Polish, get_stemmer will return a CallableStemmer that wraps a StempelStemmer. For any other language, it will return the original Stemmer.

**Comment by miso-belica** [comment]
Hi, thank you for the issue. If NLTK support is not good enough maybe it would be better to add the support you are suggesting into NLTK. WDYT?

**Comment by Manamama** [comment]
I have seen that you have stemmers in your code for Slovak, Greek etc. We had better add Polish there, instead. 

(BTW, I know next to nothing about such architecture, I have just been hacking here... )

**Comment by Manamama** [comment]
Anecdote: I have stumbled upon my own issue from years ago above because I needed to install its own on a completely new machine. I have forgotten how I had solved it. So here is a more elegant, fuller solution, written by a friendly AI:

## How to Add Polish Language Support to Sumy (Stemmer + Stopwords)

Here is a concise guide and an automated script to add full Polish language support to the `sumy` library. This solution integrates the `pystempel` library for stemming and adds a comprehensive list of Polish stopwords.

### TL;DR

To automatically add Polish language support to `sumy`, save the following script as `apply_sumy_polish_fix.sh` and run it in your terminal:

```bash
#!/bin/bash

# This script applies a patch to the sumy library to add Polish language support.

# Find the sumy installation directory
SUMY_PATH=$(python -c "import sumy; print(sumy.__path__[0])" 2>/dev/null)

if [ -z "$SUMY_PATH" ]; then
    echo "Error: sumy library not found. Please make sure it is installed."
    exit 1
fi

echo "Found sumy at: $SUMY_PATH"

# Create the polish.py stemmer file
STEMMER_FILE="$SUMY_PATH/nlp/stemmers/polish.py"
cat > "$STEMMER_FILE" << EOL
# -*- coding: utf-8 -*-

from __future__ import absolute_import
from __future__ import division, print_function, unicode_literals

from pystempel import Stemmer as PystempelStemmer


class PolishStemmer(object):
    """
    Stemmer for Polish language using pystempel.
    """
    def __init__(self):
        self._stemmer = PystempelStemmer.polimorf()

    def stem(self, word):
        stemmed_word = self._stemmer(word)
        return stemmed_word if stemmed_word is not None else word


_stemmer = PolishStemmer()


def stem_word(word):
    return _stemmer.stem(word)

EOL

echo "Created $STEMMER_FILE"

# Create the stopwords directory and download the stopwords file
STOPWORDS_DIR="$SUMY_PATH/data/stopwords"
STOPWORDS_FILE="$STOPWORDS_DIR/polish.txt"
mkdir -p "$STOPWORDS_DIR"
wget -O "$STOPWORDS_FILE" https://raw.githubusercontent.com/stopwords-iso/stopwords-pl/master/stopwords-pl.txt

if [ $? -eq 0 ]; then
    echo "Downloaded Polish stopwords to $STOPWORDS_FILE"
else
    echo "Error downloading Polish stopwords. Please check your internet connection."
    exit 1
fi

# Apply the patch to __init__.py
INIT_FILE="$SUMY_PATH/nlp/stemmers/__init__.py"

# Create the patch file
PATCH_FILE=$(mktemp)
cat > "$PATCH_FILE" << EOL
--- a/sumy/nlp/stemmers/__init__.py
+++ b/sumy/nlp/stemmers/__init__.py
@@ -5,6 +5,7 @@
 
 from .czech import stem_word as czech_stemmer
 from .ukrainian import stem_word as ukrainian_stemmer
+from .polish import stem_word as polish_stemmer
 
 from ..._compat import to_unicode
 from ...utils import normalize_language
@@ -22,6 +23,7 @@
         'japanese': null_stemmer,
         'korean': null_stemmer,
         'ukrainian': ukrainian_stemmer,
+        'polish': polish_stemmer,
     }
 
     def __init__(self, language):
EOL

patch "$INIT_FILE" "$PATCH_FILE"

if [ $? -eq 0 ]; then
    echo "Successfully patched $INIT_FILE"
    echo "Polish language support for sumy has been installed."
else
    echo "Error applying patch to $INIT_FILE. The patch may have already been applied."
fi

rm "$PATCH_FILE"
```

### The Solution Explained

The `sumy` library can be extended to support new languages by providing two key components: a **stemmer** and a list of **stopwords**.

1.  **Polish Stemmer (`pystempel`)**
    *   A new file, `sumy/nlp/stemmers/polish.py`, is created to wrap the `pystempel` library, which is a robust stemmer for the Polish language.
    *   The `sumy/nlp/stemmers/__init__.py` file is then patched to recognize and use this new stemmer when the language is set to `polish`.

2.  **Polish Stopwords**
    *   A comprehensive list of Polish stopwords is downloaded from the `stopwords-iso` GitHub repository.
    *   This list is saved as `polish.txt` in the `sumy/data/stopwords/` directory, where `sumy` automatically looks for stopwords.

This approach integrates Polish support seamlessly into `sumy`'s existing architecture, making it a permanent and robust solution.

Posted on behalf of Manamama by Gemini ♊ Core AI

**Comment by Manamama** [comment]
Re: https://github.com/miso-belica/sumy/issues/204#issuecomment-3461463939 - the idea is good and it works, but the details of the patch are wrong. I shall try to update it soon.

**Comment by miso-belica** [comment]
@Manamama you don't have to be afraid. It is actually not harder. Also Windows has no specific setup in this case. Also, the tests will be the [same as for other languages](https://github.com/miso-belica/sumy/blob/main/tests/test_tokenizers.py) and are not difficult. And, I am pretty sure AI can help you figure out 😏
